core/primitives/src/types.rs

```diff
@@ -120,14 +135,137 @@ impl SignedTransaction {
             nonce: 0,
             sender: AccountId::default(),
             receiver: AccountId::default(),
-            amount: 1,
-            method_name: String::new(),
+            amount: 0,
+            method_name: vec![],
             args: vec![],
         };
         SignedTransaction { sender_sig: DEFAULT_SIGNATURE, hash: hash_struct(&body), body }
     }
 }
 
+#[derive(Hash, Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub enum ReceiptBody {
+    NewCall(AsyncCall),
+    CallBack(CallBack),
+    Refund,
+}
+
+#[derive(Hash, Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct AsyncCall {
+    pub amount: u64,
+    pub mana: u32,
+    pub method_name: Vec<u8>,
+    pub args: Vec<u8>,
+    pub callback: Option<CallBackId>,
+}
+
+impl AsyncCall {
+    pub fn new(amount: u64, mana: u32, method_name: Vec<u8>, args: Vec<u8>) -> Self {
+        AsyncCall {
+            amount,
+            mana,
+            method_name,
+            args,
+            callback: None,
+        }
+    }
+}
+
+#[derive(Hash, Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct CallBack {
+    // callback id
+    pub id: CallBackId,
+    // results
+    pub results: Option<Vec<u8>>,
```

Comment:
If Callback here is part of ReceiptBody then you don't have all results. You have CallbackInfo (with result index) and only ONE result. Because it's generated from the other shard that just returned the Result from the execution. It doesn't have all other results, so it would return 1 result value that would be aggregated with other results on the Callback's shard before the actual callback code is executed. The callback code is executed only when all results arrive from other shards.

Reply:
Fixed in ca636d9